lisa/interrupts.py

```diff
@@ -0,0 +1,109 @@
+        return self.__str__()
+
+    @staticmethod
+    def get_interrupt_data(node: Node) -> List[Interrupt]:
+        # Run cat /proc/interrupts. The output is of the form :
+        #          CPU0       CPU1
+        # 0:         22          0  IR-IO-APIC   2-edge      timer
+        # 1:          2          0  IR-IO-APIC   1-edge      i8042
+        # ERR:        0
+        # The first column refers to the IRQ number. The next column contains
+        # number of interrupts per IRQ for each CPU in the system. The remaining
+        # column report the metadata about interrupts, including type of interrupt,
+        # device etc. This is variable for each distro.
+        # Note : Some IRQ numbers have single entry because they're not actually
+        # CPU stats, but events count belonging to the IO-APIC controller. For
+        # example, `ERR` is incremented in the case of errors in the IO-APIC bus.
+        result = (
+            node.tools[Cat].run("/proc/interrupts", sudo=True, force_run=True).stdout
+        )
+        mappings_with_header = result.splitlines(keepends=False)
+        mappings = mappings_with_header[1:]
+        assert len(mappings) > 0
+
+        interrupts = []
+        for line in mappings:
+            matched = Interrupt._interrupt_regex.fullmatch(line)
+            assert matched
+            interrupt_count = [
+                int(count) for count in matched.group("interrupt_count").split()
+            ]
+            interrupts.append(
+                Interrupt(
+                    irq_number=matched.group("irq_number"),
+                    interrupt_count=interrupt_count,
+                    metadata=matched.group("metadata"),
+                )
+            )
+        return interrupts
+
+    @staticmethod
+    def get_sum_of_interrupt_data_by_irqs(
+        node: Node, pci_slot: str
+    ) -> List[Dict[str, int]]:
+        interrupts_sum_by_irqs: List[Dict[str, int]] = []
+        interrupts = Interrupt.get_interrupt_data(node)
+        matched_interrupts = [x for x in interrupts if pci_slot in x.metadata]
+        for interrupt in matched_interrupts:
+            interrupts_sum_by_irqs.append(
+                {interrupt.irq_number: sum([int(x) for x in interrupt.interrupt_count])}
+            )
+        return interrupts_sum_by_irqs
+
+    @staticmethod
+    def get_sum_of_interrupt_data_by_cpus(node: Node, pci_slot: str) -> Dict[int, int]:
+        interrupts_sum_by_cpus: Dict[int, int] = {}
+        interrupts = Interrupt.get_interrupt_data(node)
+        matched_interrupts = [x for x in interrupts if pci_slot in x.metadata]
+        for cpu_index in range(0, len(matched_interrupts[0].interrupt_count)):
+            interrupts_sum_by_cpus[
+                cpu_index
+            ] = Interrupt.get_sum_of_interrupt_data_per_cpu(
+                matched_interrupts, cpu_index
+            )
+        return interrupts_sum_by_cpus
+
+    @staticmethod
+    def get_sum_of_interrupt_data_per_cpu(
```

Comment:
This is used internally, it doesn't need to be public.

Reply:
update